### PR TITLE
Facet count

### PR DESCRIPTION
**Pull request by MarinPostma**
#625

### CODE REVIEW

File: meilisearch-core/src/bucket_sort.rs [hunk 1]
```diff
@@ -107,6 +120,12 @@ where
         docids = Cow::Owned(intersection);
     }
 
+    result.facets = facet_count_docids.map(|f| {
+        // hardcoded value, until approximation optimization
+        result.exhaustive_facet_count = Some(true);
+        facet_count(f, &docids)
+    });
```
Comment (Kerollmops) [inline meilisearch-core/src/bucket_sort.rs:127]:
```suggestion
    if let Some(f) = facet_count_docids {
        // hardcoded value, until approximation optimization
        result.exhaustive_facet_count = Some(true);
        result.facets = Some(facet_count(f, &docids));
    }
```

File: meilisearch-core/src/bucket_sort.rs [hunk 2]
```diff
@@ -240,12 +265,18 @@ where
...
             .intersection()
             .into_set_buf();
         docids = Cow::Owned(intersection);
     }
 
+    result.facets = facet_count_docids.map(|f| {
+        // hardcoded value, until approximation optimization
+        result.exhaustive_facet_count = Some(true);
+        facet_count(f, &docids)
+    });
```
Comment (Kerollmops) [inline meilisearch-core/src/bucket_sort.rs:278]:
```suggestion
    if let Some(f) = facet_count_docids {
        // hardcoded value, until approximation optimization
        result.exhaustive_facet_count = Some(true);
        result.facets = Some(facet_count(f, &docids));
    }
```

File: meilisearch-core/src/bucket_sort.rs [hunk 3]
```diff
@@ -575,3 +608,21 @@ impl Deref for PostingsListView<'_> {
...
+        for (value, facet_docids) in doc_map {
+            let mut counter = Counter::new();
+            let op = OpBuilder::new(facet_docids.as_ref(), candidate_docids).intersection();
+            SetOperation::<DocumentId>::extend_collection(op, &mut counter);
+            count_map.insert(value, counter.0);
+        }
+        facets.insert(key, count_map);
+    }
+    facets
+}
```
Comment (Kerollmops) [inline meilisearch-core/src/bucket_sort.rs:628]:
```suggestion
fn facet_count(
    facets_docids: HashMap<String, HashMap<String, Cow<Set<DocumentId>>>>,
    candidate_docids: &Set<DocumentId>,
) -> HashMap<String, HashMap<String, usize>> {
    let mut facets_counts = HashMap::with_capacity(facets_docids.len());
    for (key, doc_map) in facet_count_docids {
        let mut count_map = HashMap::with_capacity(doc_map.len());
        for (value, facet_docids) in doc_map {
            let mut counter = Counter::new();
            let op = OpBuilder::new(facet_docids.as_ref(), candidate_docids).intersection();
            SetOperation::<DocumentId>::extend_collection(op, &mut counter);
            count_map.insert(value, counter.0);
        }
        facets_counts.insert(key, count_map);
    }
    facets_counts
}
```

File: meilisearch-core/src/query_builder.rs [hunk 4]
```diff
@@ -76,8 +87,8 @@ impl<'c, 'f, 'd, 'fa, 'i> QueryBuilder<'c, 'f, 'd, 'fa, 'i> {
         reader: &heed::RoTxn<MainT>,
         query: &str,
         range: Range<usize>,
-    ) -> MResult<(Vec<Document>, usize)> {
-        let facets_docids = match self.facets {
+    ) -> MResult<SortResult> {
+        let facets_docids = match self.facet_fitlers {
```
Comment (Kerollmops) [inline meilisearch-core/src/query_builder.rs:91]:
```suggestion
        let facets_docids = match self.facet_filters {
```

File: meilisearch-core/src/query_builder.rs [hunk 5]
```diff
@@ -1,35 +1,45 @@
...
     distinct: Option<(Box<dyn Fn(DocumentId) -> Option<u64> + 'd>, usize)>,
     timeout: Option<Duration>,
     index: &'i store::Index,
-    facets: Option<&'fa FacetFilter>,
+    facet_fitlers: Option<FacetFilter>,
+    facets: Option<Vec<(FieldId, String)>>,
 }
 
-impl<'c, 'f, 'd, 'fa, 'i> QueryBuilder<'c, 'f, 'd, 'fa, 'i> {
+impl<'c, 'f, 'd, 'i> QueryBuilder<'c, 'f, 'd, 'i> {
```
Comment (Kerollmops) [inline meilisearch-core/src/query_builder.rs:28]: You removed the `'fa` lifetime here but not everywhere, is it normal or intended?

File: meilisearch-core/src/query_builder.rs [hunk 6]
```diff
@@ -109,12 +120,30 @@ impl<'c, 'f, 'd, 'fa, 'i> QueryBuilder<'c, 'f, 'd, 'fa, 'i> {
             None => None
         };
 
+        let facet_count_docids = match self.facets {
```
Comment (Kerollmops) [inline meilisearch-core/src/query_builder.rs:123]: Can you add a comment just to explain a little bit what this block of code do?
It seems like it retireves the documents ids associated with facets that the user requested.

File: meilisearch-core/src/store/facets.rs [hunk 7]
```diff
@@ -22,6 +23,10 @@ impl Facets {
         self.facets.put(writer, &facet_key, doc_ids)
     }
 
+    pub fn field_document_ids<'txn>(&self, reader: &'txn RoTxn<MainT>, field_id: FieldId) -> ZResult<RoRange<'txn, FacetKey, CowSet<DocumentId>>> {
+        self.facets.prefix_iter(reader, &FacetKey::new(field_id, "".to_string()))
```
Comment (Kerollmops) [inline meilisearch-core/src/store/facets.rs:27]:
```suggestion
        self.facets.prefix_iter(reader, &FacetKey::new(field_id, String::new()))
```

File: meilisearch-core/src/store/mod.rs [hunk 8]
```diff
@@ -363,10 +363,10 @@ impl Index {
         QueryBuilder::new(self)
     }
 
-    pub fn query_builder_with_criteria<'c, 'f, 'd, 'fa, 'i>(
+    pub fn query_builder_with_criteria<'c, 'f, 'd, 'fa, 'i, 'q>(
```
Comment (Kerollmops) [inline meilisearch-core/src/store/mod.rs:366]: Is this new `'q` lifetime necessary?

Comment (Kerollmops) [reply meilisearch-core/src/store/mod.rs:366]: And is it normal that we keep the `'fa` one?

File: meilisearch-http/src/routes/search.rs [hunk 9]
```diff
@@ -150,3 +163,35 @@ async fn search_with_url_query(
 
     Ok(HttpResponse::Ok().json(search_builder.search(&reader)?))
 }
+
+fn prepare_facet_list(facets: &str, schema: &Schema, facet_attrs: &[FieldId]) -> Result<Vec<(FieldId, String)>, FacetCountError> {
```
Comment (Kerollmops) [inline meilisearch-http/src/routes/search.rs:167]: Could you add a little comment explaining what this function do, please?

File: meilisearch-core/src/store/mod.rs (hunk 8 above)
Comment (MarinPostma) [reply meilisearch-core/src/store/mod.rs:366]: that's right, forgot to remove them here. Funny rustc does not warn about unused lifetimes...

File: meilisearch-core/src/bucket_sort.rs [hunk 10]
```diff
@@ -575,3 +608,21 @@ impl Deref for PostingsListView<'_> {
         }
     }
 }
+
+fn facet_count(
```
Comment (Kerollmops) [inline meilisearch-core/src/bucket_sort.rs:612]: Coud you add a little comment explaining what this function do, please?

File: meilisearch-core/src/bucket_sort.rs [hunk 11]
```diff
@@ -575,3 +608,21 @@ impl Deref for PostingsListView<'_> {
...
+        for (value, facet_docids) in doc_map {
+            let mut counter = Counter::new();
+            let op = OpBuilder::new(facet_docids.as_ref(), candidate_docids).intersection();
+            SetOperation::<DocumentId>::extend_collection(op, &mut counter);
+            count_map.insert(value, counter.0);
+        }
+        facets_counts.insert(key, count_map);
+    }
+    facets_counts
+}
```
Comment (Kerollmops) [inline meilisearch-core/src/bucket_sort.rs:628]:
```suggestion
    facets_docids: HashMap<String, HashMap<String, Cow<Set<DocumentId>>>>,
    candidate_docids: &Set<DocumentId>,
) -> HashMap<String, HashMap<String, usize>> {
    let mut facets_counts = HashMap::with_capacity(facets_docids.len());
    for (key, doc_map) in facets_docids {
        let mut count_map = HashMap::with_capacity(doc_map.len());
        for (value, facet_docids) in doc_map {
            let mut counter = Counter::new();
            let op = OpBuilder::new(facet_docids.as_ref(), candidate_docids).intersection();
            SetOperation::<DocumentId>::extend_collection(op, &mut counter);
            count_map.insert(value, counter.0);
        }
        facets_counts.insert(key, count_map);
    }
    facets_counts
}
```

File: meilisearch-core/src/query_builder.rs [hunk 12]
```diff
@@ -1,35 +1,45 @@
...
-pub struct QueryBuilder<'c, 'f, 'd, 'fa, 'i> {
+pub struct QueryBuilder<'c, 'f, 'd, 'i> {
     criteria: Criteria<'c>,
     searchable_attrs: Option<ReorderedAttrs>,
     filter: Option<Box<dyn Fn(DocumentId) -> bool + 'f>>,
     distinct: Option<(Box<dyn Fn(DocumentId) -> Option<u64> + 'd>, usize)>,
     timeout: Option<Duration>,
     index: &'i store::Index,
-    facets: Option<&'fa FacetFilter>,
+    facet_filters: Option<FacetFilter>,
```
Comment (Kerollmops) [inline meilisearch-core/src/query_builder.rs:24]: Could you rename the `FcaetFilter` type into `FacetFilters` please?

File: meilisearch-core/src/query_builder.rs [hunk 13]
```diff
@@ -1,35 +1,45 @@
...
     pub fn new(index: &'i store::Index) -> Self {
         QueryBuilder::with_criteria(
             index,
             Criteria::default(),
         )
     }
 
-    pub fn set_facets(&mut self, facets: Option<&'fa FacetFilter>) {
+    /// sets facet attributes to filter on
+    pub fn set_facet_filters(&mut self, facets: Option<FacetFilter>) {
```
Comment (Kerollmops) [inline meilisearch-core/src/query_builder.rs:37]:
```suggestion
    pub fn set_facet_filter(&mut self, facets: Option<FacetFilter>) {
```

**Pull request merged**